### ISSUE TITLE
Notification lifetime is too short

### ISSUE THREAD

**Issue by CvX** [electron/electron#16922]
* **Electron Version** (output of `node_modules/.bin/electron --version`):
  * v4.0.4 / Electron 6.0.0-nightly.20190212
* **Operating System** (Platform and Version):
  * macOS 10.14.2
* **Last known working Electron version** (if applicable):
  * n/a?

### Expected Behavior

Electron's [Notification](https://electronjs.org/docs/api/notification) events should always be fired after clicking, replying, selecting an action, or closing a notification. 

### Actual behavior

As found here: https://github.com/sindresorhus/caprine/pull/726#issuecomment-457942388 Electron's [Notification](https://electronjs.org/docs/api/notification) objects are being prematurely garbage collected, so in many cases clicking, replying, or closing a notification doesn't trigger anything because the underlying Notification object (and its callbacks) is already gone.

### To Reproduce

https://github.com/CvX/electron-quick-start/tree/notifications-lifetime

```sh
$ git clone https://github.com/CvX/electron-quick-start.git -b notifications-lifetime
$ npm install
$ npm start
```

Click on "reply" in the notification, enter some text, and wait for another notification or two before submitting. You'll notice that the reply sometimes won't be logged. The event listener will execute only if click/close/reply happens _shortly_ after the notification is created.

To see what's happening under the hood use: https://github.com/CvX/electron/tree/log-notification-destroy. It will show when notifications are being destroyed, e.g.:

```sh
$ ./out/Debug/Electron.app/Contents/MacOS/Electron --enable-logging ~/dev/electron-quick-start/
(…)
show 1
[87262:0213/050539.674577:INFO:atom_api_notification.cc(82)] destroying a Notification (0x7fd5e2c9be80)
show 2
show 3
show 4
[87262:0213/050554.729886:INFO:atom_api_notification.cc(82)] destroying a Notification (0x7fd5e2ca0dd0)
[87262:0213/050554.729963:INFO:atom_api_notification.cc(82)] destroying a Notification (0x7fd5e2ca13f0)
[87262:0213/050554.730000:INFO:atom_api_notification.cc(82)] destroying a Notification (0x7fd5e2c9f800)
show 5
```

(Here, no notifications were closed yet 4 of them have been destroyed)

### Additional Information

To fix this, Electron has to keep track of notifications that have event listeners, as stated in [Notification Web API spec](https://notifications.spec.whatwg.org/#garbage-collection).

**Comment by MarshallOfSound** [comment]
When developing apps with Electron you have to be much more aware of GC than normal.  For instance you need to ensure a reference to `Tray` objects is retained or it will be GC'ed.  The same concept applies to main process Notifications, if you want it to be long lived you need to ensure that you retain a reference to the `Notification` object otherwise it will be GC'ed as we only retain a weak reference internally.

It has been discussed if we should make it a strong reference and prevent Notification's being GC'ed but it was determined that that decision should be on the app.  If you want your notification to hang around forever you should insert it into a globally scoped object or something.

```js
const notifications = {};
let id = 1;

const showNotification = () => {
  notifications[id++] = new Notification()
}
```

> To fix this, Electron has to keep track of notifications that have event listeners, as stated in [Notification Web API spec](https://notifications.spec.whatwg.org/#garbage-collection).

The Electron main process Notification API is not designed or intended to be spec compliant with the renderer side whatwg Notification API.


###############

My preference here is to document this behavior similar to the `Tray` behavior, though I am also open to revisiting whether we should leak `Notification` object to prevent this "gotcha"

**Comment by CvX** [comment]
Hey, thanks for the reply! 🙂 

> it was determined that that decision should be on the app

Wouldn't every app that uses event listeners with Notifications need to add this boilerplate code?

Otherwise even cases as simple as this can (and do) fail:

```js
const notification = new Notification(…);
notification.on('show', () => console.log('show'));
notification.show();
```

Also, Tray and Notification objects are usually used in slightly different way. Tray is most often kept for the whole lifetime of a process, while Notification is expected to be discarded as soon as it's interacted with. That's how I handled it in Caprine:

```js
const notifications = new Map();

ipc.on('notification', (…) => 
  const notification = new Notification(…);

  notifications.set(id, notification);

  notification.on('click', () => {
    …
    notifications.delete(id);
  });

  notification.on('reply', () => {
    …
    notifications.delete(id);
  });

  notification.on('close', () => {
    …
    notifications.delete(id);
  });

  notification.show();
);
```

I'd say it makes sense for Electron to take care of that.

**Comment by felixrieseberg** [comment]
Hey everyone! Thank you all so much for giving Electron a try and thinking about ways to make it better. As you can probably tell from how this issue has been open for more than five years, so far it has never been important enough for someone to close it directly or to otherwise make the resources available for it to happen.

If you're wondering what you can do to get this into Electron, great news - Electron is truly open source! We, the maintainers, [welcome outside contributions - especially from individuals and companies interested in joining us in maintaining this platform](https://www.electronjs.org/governance).

If you're reading this and are thinking to yourself "Alright, I'll do it!" - good for you! For smaller changes, you can usually open a well-tested and documented PR directly on this repository. For larger changes, I'd recommend that you start with a "[request for comments](https://www.electronjs.org/blog/rfcs)". This will ensure that you don't spend a lot of time on a larger change without getting feedback about your approach first. Alternatively, you could always hire a consultant, either from the maintainer community our outside, to perform the work that's important to you.

In short, I'm closing this issue because we're currently not planning this work, but if you care deeply about this issue, we'd welcome your contribution. For this issue in particular, it's likely that it's actually an upstream issue with Chromium, too.
